WordPress/try-wordpress
# Support for building the UX to liberate a simple post

This PR provides the support for building the UX to liberate a simple post.

- [x] REST API handling
- [x] Test preview url for preview in playground
- [x] Promote to final post

To recap the behavior, once the user has started the import, frontend would start saving the post with each selection made by user. First save would save the post as a "draft" with `liberated_post` CPT (custom post type) and then retain its ID. Consecutive saves would be update of this draft post. Details on meta fields follow after this section. Preview url for this post is what is shown in the playground. Once the user is happy with the preview & indicates extraction of this post is done, they can invoke the `promote` endpoint from a button press. We don't update the `post_status` to `publish`, it's always kept as `draft`. Essentially we save both the raw data and processed data by our "native processing" implementation. With the `promote` endpoint call, a duplicated post is created with `post_type` as `post`, which becomes the final result of liberation, id of which is saved in `_liberated_post` postmeta in `liberated_post` CPT.

## Meta fields explanation:

This PR registers meta fields on custom post types so that we can use them via REST API.

We are saving following meta fields to begin with:

- `guid` Original url of the post we are migrating
- `raw_title` Pre-processed post title as chosen/detected from frontend
- `raw_date` Pre-processed post date as chosen/detected from frontend
- `raw_content` Pre-processed post content as chosen/detected from frontend

`guid` and `raw_content` (saved in `post_content_filtered` column) both are saved in `wp_posts` table as opposed to `wp_postmeta` table, despite defined as meta fields. This is because they are not exposed on REST API directly. But both are visible under meta regardless when querying with a GET call.

The idea of saving these raw fields pre-processing, even excluding Gutenberg paste handler is for better re-transformations possibility in future.

## Test instructions

Start the environment: `wp-env start`

Create custom post type:
```
curl -X POST \
  http://localhost:8888/wp-json/wp/v2/liberated_posts \
  -H "Content-Type: application/json" \
  -d '{
    "title": "New Post from API Client",
    "content": "This is a new post created via the API client.",
    "status": "draft",
    "date_gmt": "2024-09-12T14:30:00",
    "author": 1,
    "slug": "sample-post-via-api",
    "meta": {
      "guid": "https://example.org/2222",
      "raw_title": "<h1>title</h1>",
      "raw_date": "<div>12 Dec 2024 2:30PM</div>",
      "raw_content": "<div><p>Content 1</p><p>Content 2</p></div>"
    }
  }'
```

Then check inside post and post meta with the returned ID in JSON response:
```
wp-env run cli wp post get 102
```
```
ℹ Starting 'wp post get 102' on the cli container.

+-----------------------+-------------------------------------------------------+
| Field                 | Value                                                 |
+-----------------------+-------------------------------------------------------+
| ID                    | 102                                                   |
| post_author           | 1                                                     |
| post_date             | 2024-09-12 14:30:00                                   |
| post_date_gmt         | 2024-09-12 14:30:00                                   |
| post_content          | This is a new post created via the API client.        |
| post_title            | New Post from API Client                              |
| post_excerpt          |                                                       |
| post_status           | draft                                                 |
| comment_status        | closed                                                |
| ping_status           | closed                                                |
| post_password         |                                                       |
| post_name             | sample-post-via-api-66                                |
| to_ping               |                                                       |
| pinged                |                                                       |
| post_modified         | 2024-09-12 14:30:00                                   |
| post_modified_gmt     | 2024-09-12 14:30:00                                   |
| post_content_filtered | <div><p>Content 1</p><p>Content 2</p></div>           |
| post_parent           | 0                                                     |
| guid                  | https://example.org/2222                              |
| menu_order            | 0                                                     |
| post_type             | liberated_post                                        |
| post_mime_type        |                                                       |
| comment_count         | 0                                                     |
| url                   | http://localhost:8888/?post_type=liberated_post&p=102 |
+-----------------------+-------------------------------------------------------+
✔ Ran `wp post get 102` in 'cli'. (in 1s 365ms)
```

```
wp-env run cli wp post meta list 102
```
```
ℹ Starting 'wp post meta list 102' on the cli container.

+---------+-----------+-------------------------------+
| post_id | meta_key  | meta_value                    |
+---------+-----------+-------------------------------+
| 102     | raw_title | <h1>title</h1>                |
| 102     | raw_date  | <div>12 Dec 2024 2:30PM</div> |
+---------+-----------+-------------------------------+
✔ Ran `wp post meta list 102` in 'cli'. (in 1s 277ms)
```

Promote a post:
```
curl -X POST http://localhost:8888/wp-json/wp/v2/liberated_posts/102/promote
```
```
{"success":true,"message":"Post promoted successfully","post_id":"138"}
```

A reference to promoted post is saved in postmeta:
```
wp-env run cli wp post meta get 102 _liberated_post
```
```
138
```
